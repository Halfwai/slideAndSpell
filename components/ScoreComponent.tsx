import React, { useState, useEffect, useRef } from 'react';
import { View, Text, StyleSheet, Animated, Dimensions } from 'react-native';
import { COLOURS } from '@/constants/colours';

interface ScoreComponentProps {
    slides: number;
    incrementTime: boolean;
    gameOver: boolean;
    returnTime: Function;
    slideUp: boolean;
}

export default function ScoreComponents(props: ScoreComponentProps) {
    const timeXPos = useRef(new Animated.Value(-Dimensions.get('window').width)).current;
    const slidesXPos = useRef(new Animated.Value(Dimensions.get('window').width)).current;
    const slideYPos = useRef(new Animated.Value(200)).current;

    const incrementTimeRef = useRef(props.incrementTime);

    useEffect(() => {
        incrementTimeRef.current = props.incrementTime;
    }, [props.incrementTime]);

    useEffect(() => {
        if (props.slideUp) {
            animateSlideUp();
        }
    }, [props.slideUp]);

    useEffect(() => {
        if(props.gameOver) {
            props.returnTime(time);
        }
    }, [props.gameOver]);

    const [time, setTime] = useState(0);

    function increment() {
        if (!incrementTimeRef.current) return;
        setTime((time) => time + 1);
    }

    function returnTimeString() {
        const seconds = time % 60;
        const minutes = Math.floor(time / 60) % 60;
        const hours = Math.floor(time / 3600);
        return `${hours < 10 ? "0" : ""}${hours}:${minutes < 10 ? "0" : ""}${minutes}:${seconds < 10 ? "0" : ""}${seconds}`;
    }

    // increment time function adapted from here - https://iq.js.org/questions/react/how-to-update-a-component-every-second
    useEffect(() => {
        const interval = setInterval(() => {
            increment();
        }, 1000);
        return () => clearInterval(interval);
    }, []);

    useEffect(() => {
        Animated.timing(timeXPos, {
            toValue: 0,
            duration: 200,
            useNativeDriver: true
        }).start();
        Animated.timing(slidesXPos, {
            toValue: 0,
            duration: 200,
            useNativeDriver: true
        }).start();
    }, []);

    function animateSlideUp() {
        Animated.timing(slideYPos, {
            toValue: -200,
            duration: 200,
            useNativeDriver: true
        }).start();
    }


    return (
        <View style={styles.container}>
            <Animated.View style={[styles.itemBox, { transform: [{ translateX: timeXPos }, { translateY: slideYPos }] }]}>

                    <Text style={styles.itemText}>Time</Text>
                    <Text style={styles.itemText}>{returnTimeString()}</Text>

            </Animated.View>
            <Animated.View style={[styles.itemBox, { transform: [{ translateX: slidesXPos }, { translateY: slideYPos }] }]}>

                    <Text style={styles.itemText}>Slides</Text>
                    <Text style={styles.itemText}>{props.slides}</Text>
            </Animated.View>
        </View>


    )
};

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        marginTop: 20,
        zIndex: 0,
        position: 'absolute',
    },
    itemBox: {
        flexDirection: 'column', 
        justifyContent: 'center', 
        alignItems: 'center',
        width: Dimensions.get('window').width / 2,
        backgroundColor: COLOURS.green,
        paddingVertical: 10,
        borderTopWidth: 5,
        borderBottomWidth: 5,
        borderColor: 'white',
    },
    itemText: {
        fontSize: 20,
        color: 'black',
        textAlign: 'center'
    }   
});