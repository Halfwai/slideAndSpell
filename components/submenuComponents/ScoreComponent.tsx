import React, { useState, useEffect, useRef } from 'react';
import { View, Text, StyleSheet, Animated, Dimensions } from 'react-native';

// Import COLOURS
import { COLOURS } from '@/constants/colours';
import { SIZES } from '@/constants/sizes';

// Setup props
interface ScoreComponentProps {
    slides: number;
    incrementTime: boolean;
    gameOver: boolean;
    returnTime: Function;
    slideUp: boolean;
}

export default function ScoreComponents(props: ScoreComponentProps) {
    // Setup animation values
    const timeXPos = useRef(new Animated.Value(-Dimensions.get('window').width)).current;
    const slidesXPos = useRef(new Animated.Value(Dimensions.get('window').width)).current;
    const slideYPos = useRef(new Animated.Value(SIZES.GAMEBOARD / 1.5)).current;
    // Setup ref for incrementTime
    const incrementTimeRef = useRef(props.incrementTime);

    // Update incrementTimeRef when props.incrementTime changes
    useEffect(() => {
        incrementTimeRef.current = props.incrementTime;
    }, [props.incrementTime]);

    // Slide up animation
    useEffect(() => {
        if (props.slideUp) {
            animateSlideUp();
        }
    }, [props.slideUp]);

    // Return time when game is over
    useEffect(() => {
        if (props.gameOver) {
            props.returnTime(time);
        }
    }, [props.gameOver]);

    // Setup time state
    const [time, setTime] = useState(0);

    // Increment time function
    function increment() {
        if (!incrementTimeRef.current) return;
        setTime((time) => time + 1);
    }

    // Formats the time into a readable string
    function returnTimeString() {
        const seconds = time % 60;
        const minutes = Math.floor(time / 60) % 60;
        const hours = Math.floor(time / 3600);
        return `${hours < 10 ? "0" : ""}${hours}:${minutes < 10 ? "0" : ""}${minutes}:${seconds < 10 ? "0" : ""}${seconds}`;
    }

    // increment time function adapted from here - https://iq.js.org/questions/react/how-to-update-a-component-every-second
    useEffect(() => {
        const interval = setInterval(() => {
            increment();
        }, 1000);
        return () => clearInterval(interval);
    }, []);

    // Slide score components in each side
    useEffect(() => {
        Animated.timing(timeXPos, {
            toValue: 0,
            duration: 200,
            useNativeDriver: true
        }).start();
        Animated.timing(slidesXPos, {
            toValue: 0,
            duration: 200,
            useNativeDriver: true
        }).start();
    }, []);

    // Slide up animation
    function animateSlideUp() {
        Animated.timing(slideYPos, {
            toValue: -Dimensions.get('window').height * 0.44,
            duration: 200,
            useNativeDriver: true
        }).start();
    }


    return (
        <View style={styles.container}>
            <Animated.View style={[styles.itemBox, { transform: [{ translateX: timeXPos }, { translateY: slideYPos }] }]}>
                <Text style={styles.itemText}>Time</Text>
                <Text style={styles.itemText}>{returnTimeString()}</Text>
            </Animated.View>
            <Animated.View style={[styles.itemBox, { transform: [{ translateX: slidesXPos }, { translateY: slideYPos }] }]}>
                <Text style={styles.itemText}>Slides</Text>
                <Text style={styles.itemText}>{props.slides}</Text>
            </Animated.View>
        </View>
    )
};

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        marginTop: 20,
        zIndex: 0,
        position: 'absolute',
    },
    itemBox: {
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center',
        width: Dimensions.get('window').width / 2,
        backgroundColor: COLOURS.green,
        paddingVertical: 10,
        borderTopWidth: 5,
        borderBottomWidth: 5,
        borderColor: 'white',
    },
    itemText: {
        fontSize: 20,
        color: 'black',
        textAlign: 'center'
    }
});