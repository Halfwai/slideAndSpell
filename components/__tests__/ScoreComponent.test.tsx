import { render } from '@testing-library/react-native';
import ScoreComponent from '@/components/submenuComponents/ScoreComponent';

it(`ScoreComponent Component renders correctly`, () => {
    const { toJSON } = render(
        <ScoreComponent
            slides={0}
            incrementTime={false}
            gameOver={false}
            returnTime={() => {}}
            slideUp={false}
        />
    );    
    expect(toJSON()).toMatchSnapshot();
});