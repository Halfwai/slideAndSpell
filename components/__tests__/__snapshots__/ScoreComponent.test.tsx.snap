// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreComponent Component renders correctly 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "marginTop": 20,
      "position": "absolute",
      "zIndex": 0,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#8BF0BC",
        "borderBottomWidth": 5,
        "borderColor": "white",
        "borderTopWidth": 5,
        "flexDirection": "column",
        "justifyContent": "center",
        "paddingVertical": 10,
        "transform": [
          {
            "translateX": -750,
          },
          {
            "translateY": 200,
          },
        ],
        "width": 375,
      }
    }
  >
    <Text
      style={
        {
          "color": "black",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      Time
    </Text>
    <Text
      style={
        {
          "color": "black",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      00:00:00
    </Text>
  </View>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#8BF0BC",
        "borderBottomWidth": 5,
        "borderColor": "white",
        "borderTopWidth": 5,
        "flexDirection": "column",
        "justifyContent": "center",
        "paddingVertical": 10,
        "transform": [
          {
            "translateX": 750,
          },
          {
            "translateY": 200,
          },
        ],
        "width": 375,
      }
    }
  >
    <Text
      style={
        {
          "color": "black",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      Slides
    </Text>
    <Text
      style={
        {
          "color": "black",
          "fontSize": 20,
          "textAlign": "center",
        }
      }
    >
      0
    </Text>
  </View>
</View>
`;
